File: pdf_to_latex_batch.py
#!/usr/bin/env python3
"""
PDF to LaTeX Batch Converter
Converts all PDFs in a folder to LaTeX format using Claude API
"""

import os
import sys
import json
import subprocess
from pathlib import Path
import anthropic

def extract_text_from_pdf(pdf_path: str) -> str:
    """Extract text from PDF using pdftotext"""
    try:
        result = subprocess.run(
            ['pdftotext', '-layout', pdf_path, '-'],
            capture_output=True,
            text=True,
            timeout=30
        )
        if result.returncode == 0:
            return result.stdout
        else:
            print(f"⚠️  pdftotext error for {pdf_path}: {result.stderr}")
            return ""
    except FileNotFoundError:
        print("❌ pdftotext not found. Install with: apt-get install poppler-utils")
        sys.exit(1)
    except subprocess.TimeoutExpired:
        print(f"⏱️  Timeout extracting text from {pdf_path}")
        return ""

def convert_to_latex(pdf_text: str, filename: str) -> str:
    """Use Claude to convert extracted PDF text to LaTeX"""
    client = anthropic.Anthropic()
    
    message = client.messages.create(
        model="claude-opus-4-7",
        max_tokens=4096,
        messages=[
            {
                "role": "user",
                "content": f"""Convert the following PDF content to a well-formatted LaTeX document.

Guidelines:
- Use proper LaTeX document structure with \\documentclass{{article}}
- Convert headers to \\section{{}} and \\subsection{{}}
- Convert bullet points to itemize environments
- Convert tables to tabular environments
- Convert mathematical notation to proper LaTeX math mode ($ $ or \[ \])
- Preserve code blocks in verbatim or lstlisting environments
- Add appropriate packages in preamble (amsmath, graphicx, etc. as needed)
- Keep the content structure and meaning
- Use proper LaTeX formatting throughout

PDF Content:
---
{pdf_text[:10000]}  # Limit to first 10k chars to avoid huge context
---

Output only the complete LaTeX code, ready to compile."""
            }
        ]
    )
    
    return message.content[0].text

def process_pdf_folder(folder_path: str, output_folder: str = None):
    """Process all PDFs in a folder"""
    
    # Set output folder
    if output_folder is None:
        output_folder = os.path.join(folder_path, "latex_output")
    
    # Create output directory
    Path(output_folder).mkdir(parents=True, exist_ok=True)
    
    # Find all PDF files
    pdf_files = list(Path(folder_path).glob("*.pdf"))
    
    if not pdf_files:
        print(f"❌ No PDF files found in {folder_path}")
        return
    
    print(f"✅ Found {len(pdf_files)} PDF file(s)")
    print(f"📁 Output folder: {output_folder}\n")
    
    # Process each PDF
    for idx, pdf_path in enumerate(pdf_files, 1):
        print(f"[{idx}/{len(pdf_files)}] Processing: {pdf_path.name}")
        
        # Step 1: Extract text from PDF
        print("  → Extracting text from PDF...")
        pdf_text = extract_text_from_pdf(str(pdf_path))
        
        if not pdf_text:
            print(f"  ⚠️  Skipping {pdf_path.name} - No text extracted")
            continue
        
        # Step 2: Convert to LaTeX using Claude
        print("  → Converting to LaTeX with Claude...")
        try:
            latex_code = convert_to_latex(pdf_text, pdf_path.name)
            
            # Step 3: Save LaTeX file
            output_filename = pdf_path.stem + ".tex"
            output_path = os.path.join(output_folder, output_filename)
            
            with open(output_path, 'w', encoding='utf-8') as f:
                f.write(latex_code)
            
            print(f"  ✅ Saved: {output_filename}")
            
        except Exception as e:
            print(f"  ❌ Error converting {pdf_path.name}: {str(e)}")
    
    print(f"\n✨ Conversion complete! Check {output_folder} for LaTeX files.")

def main():
    """Main entry point"""
    if len(sys.argv) < 2:
        print("Usage: python pdf_to_latex_batch.py <pdf_folder> [output_folder]")
        print("\nExample:")
        print("  python pdf_to_latex_batch.py ./pdfs ./latex_files")
        sys.exit(1)
    
    pdf_folder = sys.argv[1]
    output_folder = sys.argv[2] if len(sys.argv) > 2 else None
    
    # Validate input folder
    if not os.path.isdir(pdf_folder):
        print(f"❌ Folder not found: {pdf_folder}")
        sys.exit(1)
    
    # Check for API key
    if not os.getenv('ANTHROPIC_API_KEY'):
        print("❌ ANTHROPIC_API_KEY environment variable not set")
        print("Set it with: export ANTHROPIC_API_KEY='your-api-key'")
        sys.exit(1)
    
    # Process PDFs
    process_pdf_folder(pdf_folder, output_folder)

if __name__ == "__main__":
    main()

File: pdf_to_latex_batch.js
#!/usr/bin/env node
/**
 * PDF to LaTeX Batch Converter for Claude Code
 * Uses Claude API to convert multiple PDFs to LaTeX format
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const Anthropic = require('@anthropic-ai/sdk');

const client = new Anthropic();

/**
 * Extract text from PDF using pdftotext
 */
function extractTextFromPdf(pdfPath) {
  try {
    const text = execSync(`pdftotext -layout "${pdfPath}" -`, {
      encoding: 'utf-8',
      timeout: 30000,
    });
    return text;
  } catch (error) {
    if (error.code === 'ENOENT') {
      console.error('❌ pdftotext not found. Install with: apt-get install poppler-utils');
      process.exit(1);
    }
    console.warn(`⚠️  Error extracting text from ${pdfPath}: ${error.message}`);
    return '';
  }
}

/**
 * Convert PDF text to LaTeX using Claude
 */
async function convertToLatex(pdfText, filename) {
  const message = await client.messages.create({
    model: 'claude-opus-4-7',
    max_tokens: 4096,
    messages: [
      {
        role: 'user',
        content: `Convert the following PDF content to a well-formatted LaTeX document.

Guidelines:
- Use proper LaTeX document structure with \\documentclass{article}
- Convert headers to \\section{} and \\subsection{}
- Convert bullet points to itemize environments
- Convert tables to tabular environments
- Convert mathematical notation to proper LaTeX math mode ($ $ or \\[ \\])
- Preserve code blocks in verbatim or lstlisting environments
- Add appropriate packages in preamble (amsmath, graphicx, etc. as needed)
- Keep the content structure and meaning
- Use proper LaTeX formatting throughout

PDF Content:
---
${pdfText.substring(0, 10000)}
---

Output only the complete LaTeX code, ready to compile.`,
      },
    ],
  });

  return message.content[0].text;
}

/**
 * Process all PDFs in a folder
 */
async function processPdfFolder(folderPath, outputFolder = null) {
  // Set output folder
  if (!outputFolder) {
    outputFolder = path.join(folderPath, 'latex_output');
  }

  // Create output directory
  if (!fs.existsSync(outputFolder)) {
    fs.mkdirSync(outputFolder, { recursive: true });
  }

  // Find all PDF files
  const files = fs.readdirSync(folderPath);
  const pdfFiles = files.filter(file => file.toLowerCase().endsWith('.pdf'));

  if (pdfFiles.length === 0) {
    console.log(`❌ No PDF files found in ${folderPath}`);
    return;
  }

  console.log(`✅ Found ${pdfFiles.length} PDF file(s)`);
  console.log(`📁 Output folder: ${outputFolder}\n`);

  // Process each PDF
  for (let idx = 0; idx < pdfFiles.length; idx++) {
    const filename = pdfFiles[idx];
    const pdfPath = path.join(folderPath, filename);

    console.log(`[${idx + 1}/${pdfFiles.length}] Processing: ${filename}`);

    try {
      // Step 1: Extract text from PDF
      console.log('  → Extracting text from PDF...');
      const pdfText = extractTextFromPdf(pdfPath);

      if (!pdfText) {
        console.log(`  ⚠️  Skipping ${filename} - No text extracted`);
        continue;
      }

      // Step 2: Convert to LaTeX using Claude
      console.log('  → Converting to LaTeX with Claude...');
      const latexCode = await convertToLatex(pdfText, filename);

      // Step 3: Save LaTeX file
      const outputFilename = path.basename(filename, '.pdf') + '.tex';
      const outputPath = path.join(outputFolder, outputFilename);

      fs.writeFileSync(outputPath, latexCode, 'utf-8');
      console.log(`  ✅ Saved: ${outputFilename}\n`);
    } catch (error) {
      console.error(`  ❌ Error processing ${filename}: ${error.message}\n`);
    }
  }

  console.log(`✨ Conversion complete! Check ${outputFolder} for LaTeX files.`);
}

/**
 * Main entry point
 */
async function main() {
  const args = process.argv.slice(2);

  if (args.length < 1) {
    console.log('Usage: node pdf_to_latex_batch.js <pdf_folder> [output_folder]');
    console.log('\nExample:');
    console.log('  node pdf_to_latex_batch.js ./pdfs ./latex_files');
    process.exit(1);
  }

  const pdfFolder = args[0];
  const outputFolder = args[1] || null;

  // Validate input folder
  if (!fs.existsSync(pdfFolder) || !fs.statSync(pdfFolder).isDirectory()) {
    console.error(`❌ Folder not found: ${pdfFolder}`);
    process.exit(1);
  }

  // Check for API key
  if (!process.env.ANTHROPIC_API_KEY) {
    console.error('❌ ANTHROPIC_API_KEY environment variable not set');
    console.error("Set it with: export ANTHROPIC_API_KEY='your-api-key'");
    process.exit(1);
  }

  // Process PDFs
  await processPdfFolder(pdfFolder, outputFolder);
}

main().catch(error => {
  console.error('Fatal error:', error);
  process.exit(1);
});
